Spatial_Engine/Query/IsFullyOnGrid.cs

```diff
@@ -0,0 +1,132 @@
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.Engine.Geometry;
+using BH.oM.Dimensional;
+using BH.oM.Geometry;
+using BH.oM.Quantities.Attributes;
+using BH.oM.Reflection.Attributes;
+using BH.oM.Geometry.SettingOut;
+using System.Collections.Generic;
+using System.ComponentModel;
+
+namespace BH.Engine.Spatial
+{
+    public static partial class Query
+    {
+        /******************************************/
+        /****  Public Methods                  ****/
+        /******************************************/
+
+        [Description("Checks if the geometrical representation of an IElement0D, projected onto the XY Plane, is fully within a set tolerance from a grid line.")]
+        [Input("element0D", "The IElement0D that will be checked for proximity to the grid.")]
+        [Input("grid", "The Grid to use for evaulation.")]
+        [Input("tolerance", "The maximum distance allowed from the Grid for this method to return true.", typeof(Length))]
+        [Output("isFullyOnGrid", "A boolean which is true if the geometrical representation of the IElement0D is fully within a set tolerance from the grid line.")]
+        public static bool IsFullyOnGrid(this IElement0D element0D, Grid grid, double tolerance = BH.oM.Geometry.Tolerance.Distance)
+        {
+            Point position = element0D.IGeometry().Project(Plane.XY);
+            ICurve gridCurve = grid.Curve.IProject(Plane.XY);
+
+            return position.IDistance(gridCurve) <= tolerance;
+        }
+
+        /******************************************/
+
+        [Description("Checks if the geometrical representation of an IElement1D, projected onto the XY Plane, is fully within a set tolerance from a grid line.")]
+        [Input("element1D", "The IElement1D that will be checked for proximity to the grid.")]
+        [Input("grid", "The Grid to use for evaulation.")]
+        [Input("tolerance", "The maximum distance allowed from the Grid for this method to return true.", typeof(Length))]
+        [Output("isFullyOnGrid", "A boolean which is true if the geometrical representation of an IElement1D is fully within a set tolerance from the grid line.")]
+        public static bool IsFullyOnGrid(this IElement1D element1D, Grid grid, double tolerance = BH.oM.Geometry.Tolerance.Distance)
+        {
+            ICurve curve = element1D.IGeometry().IProject(Plane.XY);
+            List <Point> ctrlPts = curve.ControlPoints();
```

Comment:
Think we are better off projecting the controlpoints rather than the curve, as projecting the curve actually might turn it into a nurbs curve. (arcs and circles for example). Could help with your check below.

This is, ofc, unless you had it thew way you had it before intentionally

```suggestion
            ICurve curve = element1D.IGeometry();
            List <Point> ctrlPts = curve.ControlPoints().Select(x => x.Project(Plane.XY)).ToList();
```

Can also add, though, that checking the control points might not be fully rigorous (arc vs a polyline for example, might give you a false "true", but lets catch up on that next week.)

Reply:
That does sound like a good idea. There's no particular reason why the curve is projected rather than the points.

You're right about the potential for false positives. I initially had the idea of using gradient descent to find local maximi in distance, but I figured it might be too heavy and complex for a method like this. I'm sure there are many more ways of making better algorithms than what is there currently. Let's catch up on that.

Reply:
I got rid of this check.